### ISSUE TITLE
Crash: libjvm.so unnecessary_acquire

### ISSUE THREAD

**Issue by fvasco** [corretto/corretto-21#46]
Hello,
we got the following crash:

~~~
Current thread (0x0000ffff801f7df0):  JavaThread "C2 CompilerThread0" daemon [_thread_in_native, id=37, stack(0x0000ffff846b7000,0x0000ffff848b5000) (2040K)]


Current CompileTask:
C2:  58449 13223       4       kotlinx.coroutines.scheduling.CoroutineScheduler$Worker::findTask (18 bytes)

Stack: [0x0000ffff846b7000,0x0000ffff848b5000],  sp=0x0000ffff848b0d00,  free space=2023k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x2ff0bc]  unnecessary_acquire(Node const*)+0x3c
V  [libjvm.so+0x6683b8]  State::DFA(int, Node const*)+0x1d40
V  [libjvm.so+0xad0384]  Matcher::Label_Root(Node const*, State*, Node*, Node*&)+0x3f4
V  [libjvm.so+0xad0f58]  Matcher::match_tree(Node const*)+0xa8
V  [libjvm.so+0xad5004]  Matcher::xform(Node*, int)+0x77c
V  [libjvm.so+0xad7ae8]  Matcher::match()+0x740
V  [libjvm.so+0x5b601c]  Compile::Code_Gen()+0x9c
V  [libjvm.so+0x5b9528]  Compile::Compile(ciEnv*, ciMethod*, int, Options, DirectiveSet*)+0x10c8
V  [libjvm.so+0x4f1d74]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x114
V  [libjvm.so+0x5be800]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0x9b0
V  [libjvm.so+0x5c158c]  CompileBroker::compiler_thread_loop()+0x574
V  [libjvm.so+0x814bdc]  JavaThread::thread_main_inner() [clone .part.0]+0xa4
V  [libjvm.so+0xd87d68]  Thread::call_run()+0xa8
V  [libjvm.so+0xb7d7cc]  thread_native_entry(Thread*)+0xdc
C  [libc.so.6+0x7ee58]

siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x0000000000000000
~~~

we attach logs for details.

The class `kotlinx.coroutines.scheduling.CoroutineScheduler$Worker` is in the maven package [`org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm:1.7.3`](https://mvnrepository.com/artifact/org.jetbrains.kotlinx/kotlinx-coroutines-core-jvm/1.7.3).

After this error, server crash. There is a workaround?

Logs:

- [hs_err_pid24.log](https://github.com/corretto/corretto-21/files/14176693/hs_err_pid24.log)
- [replay_pid24.log](https://github.com/corretto/corretto-21/files/14176694/replay_pid24.log)

**Comment by fvasco** [comment]
I looks a duplicate of the https://github.com/corretto/corretto-17/issues/93
I confirm that this issue happened on an EC2 t4g.xlarge with Shenandoah GC.

**Comment by navyxliu** [comment]
hi, @fvasco , 
Thank you for your hs_err file and replay file. so far, what we know is that there's something wrong in C2 compiler when java is running on aarch64 with shenandoah gc. 

One quick workaround is not to use Shenandoah gc.  I will see if you can trigger this problem using your replay file. if we can reproduce this issue on our side, we will looking into the rootcause.

**Comment by fvasco** [comment]
Thank you @navyxliu,
Shenandoah is the only low pause GC that support Compressed Oops (Java Heap < 32GiB), but I confirm you that avoiding Graviton is a good workaround.
We will await for fix.

**Comment by fvasco** [comment]
I tried to create a reproducer.

The code is:

```kotlin
fun main() {
    runBlocking {
        val consumerCount = 1_000
        val messageCount = 1000

        repeat(50) {
            val mutableStateFlow = MutableSharedFlow<Int>()
            val sharedFlow = mutableStateFlow.asSharedFlow()
            val nanos: Long
            coroutineScope {
                repeat(consumerCount) {
                    launch(start = CoroutineStart.UNDISPATCHED) {
                        val channel = sharedFlow.produceIn(this)
                        repeat(messageCount) { channel.receive() }
                        cancel()
                    }
                }

                delay(1.seconds)
                nanos = System.nanoTime()
                launch(Dispatchers.Default) {
                    repeat(messageCount) { mutableStateFlow.emit(it) }
                }
            }

            val delta = System.nanoTime() - nanos
            println(NumberFormat.getIntegerInstance().format(delta / consumerCount / messageCount))
        }
    }
}
```

I did some benchmark with different counts

~~~
val consumerCount = 1_000
val messageCount = 100
824
794
819
823

val consumerCount = 10_000
val messageCount = 100
11.109
11.145
11.019

val consumerCount = 1_000
val messageCount = 1_000
773
784
758

val consumerCount = 10_000
val messageCount = 1_000
6.010
10.349
10.427
10.556

val consumerCount = 10_000
val messageCount = 10_000
10.623
~~~

I attach the JFR
[reproducer.zip](https://github.com/corretto/corretto-21/files/14195579/reproducer.zip)

**Comment by lutkerd** [comment]
It looks like the issue was also reported upstream and a fix was pushed: https://bugs.openjdk.org/browse/JDK-8325372. We will be monitoring this change and backport it once it is proven to be stable.

**Comment by lutkerd** [comment]
This issue should be resolved in [Corretto 21.0.3.9.1](https://github.com/corretto/corretto-21/releases/tag/21.0.3.9.1). Please update and let us know if you still see the issue.

Thanks.
